lib/mix/lib/mix/tasks/test.ex

```diff
@@ -679,13 +692,36 @@ defmodule Mix.Tasks.Test do
     end
   end
 
-  defp display_warn_test_pattern(test_files, test_pattern, matched_test_files, warn_test_pattern) do
-    files = Mix.Utils.extract_files(test_files, warn_test_pattern) -- matched_test_files
+  defp display_warn_test_pattern(
+         test_files,
+         test_pattern,
+         matched_test_files,
+         warn_test_pattern,
+         warn_test_ignore_pattern,
+         elixirc_paths
+       ) do
+    # any test support files should be ignored
+    compiled_files = Mix.Utils.extract_files(elixirc_paths, [:ex])
```

Comment:
I am worried about this additional extraction because in large projects this will return tens of thousand of files, which will add one second or more on each `mix test` invocation.

Even today's solution, of doing one additional traversal on `warn_test_pattern` is non-optimal. Ideally, we should probably do this:

1. Have one pattern for finding both unwanted and wanted test files, looking up the file system once
2. Then have a string or a regex that finds the ones we want to warn on

Then by default we could warn on all .ex files that end with `_test.ex` or any `.exs` that does not match `_helper.exs` Do you think we could adapt this new approach with what you want to achieve? 

Another option is to only do this check if a flag is given, which folks could enable on CI.

Reply:
I was worried about this too. One simple solution here would be to just perform a String.starts_with check with the elixirc_paths, which won't be 100% correct, but probably fine. If we want to only do extract_files once, I think we can only make it work if we are fine with breaking changes (e.g. not supporting glob path patterns in the config).